crates/turborepo-scm/src/package_deps.rs

```diff
@@ -1,13 +1,209 @@
 
 use crate::{hash_object::hash_objects, Error, Git, SCM};
 
 pub type GitHashes = HashMap<RelativeUnixPathBuf, String>;
 
+/// Precalculates workspace data where it makes sense to reduce overhead from
+/// lazily calculating it on demand.
+pub enum CachedPackageFileHasher<'a> {
+    Manual,
+    Git(GitCachedPackageFileHasher<'a>),
+}
+
+impl<'a> CachedPackageFileHasher<'a> {
+    /// When creating a new CachedPackageFileHasher with the git hashes, we
+    /// precalculate and store a datastructure that maps package paths to
+    /// hashes of the files in that package, and a list of files that
+    /// changed. This saves having to call out to git for each package.
+    pub fn new(
+        scm: &'a SCM,
+        repo_root: &AbsoluteSystemPath,
+        package_roots: impl Iterator<Item = &'a AnchoredSystemPath>,
+    ) -> Self {
+        match scm {
+            SCM::Git(git) => {
+                let mut map: HashMap<&AnchoredSystemPath, (GitHashes, Vec<_>)> = Default::default();
+
+                let j1 = {
+                    let git = git.clone();
+                    let repo_root = repo_root.to_owned();
+                    std::thread::spawn(move || git.git_ls_tree(&repo_root).unwrap())
+                };
+
+                let j2 = {
+                    let git = git.clone();
+                    let repo_root = repo_root.to_owned();
+                    std::thread::spawn(move || {
+                        git.append_git_status(&repo_root, &Default::default())
+                            .unwrap()
+                    })
+                };
+
+                let (path_hashes, to_hash) = {
+                    let mut path_hashes = j1.join().unwrap();
+                    let (to_hash, to_remove) = j2.join().unwrap();
+
+                    for path in to_remove {
+                        path_hashes.remove(&path);
+                    }
+                    (path_hashes, to_hash)
+                };
+
+                let span = tracing::debug_span!("constructing trie");
+                let enter = span.enter();
+
+                let mut hash_trie = qp_trie::Trie::new();
+                for (path, hash) in path_hashes {
+                    hash_trie.insert_str(path.as_str(), hash);
+                }
+
+                let mut file_trie = qp_trie::Trie::new();
+                for path in &to_hash {
+                    file_trie.insert_str(path.as_str(), ());
+                }
+
+                drop(enter);
+
+                let span = tracing::debug_span!("constructing map");
+                let enter = span.enter();
+
+                // a subpackage will always have a longer length than its parent,
+                // so we sort by length to ensure that we always and drain subpackages
+                // before their parents
+                for package in package_roots {
+                    let package_str = package.as_str();
+
+                    let (hashes, files) = map.entry(package).or_default();
```

Comment:
Just asking due to the use of `entry`, but do we ever expect a `package` to appear twice in `package_roots`? If so I believe `files` will contain two duplicate entries. Maybe just use an `map.insert` or dedupe `package_roots` before we iterate through them.